### ISSUE TITLE
Potential performance regression: Process.GetCurrentProcess() on Windows

### ISSUE THREAD

**Issue by adamsitnik** [dotnet/corefx#39761]
It looks like `Process.GetCurrentProcess()` have regressed on Windows.

```cmd
git clone https://github.com/dotnet/performance.git
cd performance
# if you don't have cli installed and want python script to download the latest cli for you
py .\scripts\benchmarks_ci.py -f netcoreapp2.2 netcoreapp3.0 --filter System.Diagnostics.Perf_Process.GetCurrentProcess
# if you do
dotnet run -p .\src\benchmarks\micro\MicroBenchmarks.csproj -c Release -f netcoreapp2.2 --runtimes netcoreapp2.2 netcoreapp3.0 --filter System.Diagnostics.Perf_Process.GetCurrentProcess
```

It's also suprising that the method is 4 times slower on Ubuntu 18 compared to Ubuntu 16.

## System.Diagnostics.Perf_Process.GetCurrentProcess

| conclusion |   Base |   Diff | Base/Diff | Modality | Operating System     | Arch  | Processor Name                              | Base Runtime    | Diff Runtime                     |
| ---------- | ------:| ------:| ---------:| --------:| -------------------- | ----- | ------------------------------------------- | --------------- | --------------------------------- |
| Same       | 445.90 | 459.75 |      0.97 |          | ubuntu 18.04         | 64bit | Intel Xeon CPU E5-1650 v4 3.60GHz           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Slower     |  85.47 | 105.94 |      0.81 |          | Windows 10.0.18362   | 64bit | Intel Xeon CPU E5-1650 v4 3.60GHz           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Slower     | 165.85 | 202.40 |      0.82 |          | ubuntu 16.04         | 64bit | Intel Xeon CPU E5-2673 v4 2.30GHz           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-01|
| Same       | 792.90 | 812.41 |      0.98 |          | ubuntu 18.04         | 64bit | Intel Xeon CPU E5-2673 v4 2.30GHz           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-01|
| Slower     | 119.86 | 130.78 |      0.92 |          | macOS Mojave 10.14.5 | 64bit | Intel Core i7-5557U CPU 3.10GHz (Broadwell) | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Slower     | 100.94 | 127.77 |      0.79 |          | Windows 10.0.18362   | 64bit | Intel Core i7-5557U CPU 3.10GHz (Broadwell) | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Slower     |  77.25 |  97.51 |      0.79 |          | Windows 10.0.18362   | 64bit | Intel Core i7-7700 CPU 3.60GHz (Kaby Lake)  | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Slower     |  94.54 | 106.37 |      0.89 |          | Windows 10.0.18362   | 64bit | AMD Ryzen 7 1800X                           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Slower     |  91.95 | 107.41 |      0.86 |          | Windows 10.0.18362   | 32bit | Intel Xeon CPU E5-1650 v4 3.60GHz           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|

/cc @danmosemsft @billwert @DrewScoggins

**Comment by krwq** [comment]
I don't believe this is 81bd671 - I can see that causing tiny regression due to reading parent on Linux but not on Windows - it only adds new API.

Looking at the code there is not much to regress there, basically all the code are those 3 functions:
- [GetCurrentProcess](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs#L1070) which only calls constructor and GetCurrentProccessId
- [GetCurrentProcessId](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs#L373) - which essentially only calls WinAPI
- [ctor](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs#L106) which only does couple of tiny assignments

I don't see any changes to any of those 3. Possibly something in Windows or coreclr?

Besides that why do we care so much about the perf of this method, it can be called once per process execution time if someone needs perf - is there any actual scenario where it would be useful for this to be super fast?

**Comment by billwert** [comment]
> Besides that why do we care so much about the perf of this method, it can be called once per process execution time if someone needs perf

Something like `Log($"{Process.GetCurrentProcess().whatver")` is a shockingly common pattern.

**Comment by krwq** [comment]
Ok, the only thing here which I see can be optimized is GetCurrentProcessId (I'm not sure if it's safe to cache entire Process instance - I suspect no) - I have locally tried caching returned value but I'm getting relatively the same results (I sanity checked with sleep to make sure my changes take effect). Considering the total runtime on my box shows as around 115ns I feel like it would make more sense to optimize ".whatever" rather than this call and comparing to printing/writing this improvement wouldn't even be noticable.

**Comment by adamsitnik** [comment]
I've profiled the code and: `Process.GetCurrentProcess()` itself has not regressed, `Process.GetCurrentProcess().Dispose()` did.

![obraz](https://user-images.githubusercontent.com/6011991/63689655-ce6d2b80-c80a-11e9-9a09-7775e015887d.png)

https://github.com/dotnet/corefx/pull/36333 have made the lock unconditional

https://github.com/dotnet/corefx/blob/61a6264018989e06d314b68d57bd91b4b72e4360/src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs#L837-L844

So now anytime we want to `Dispose` a `Process` instance, the lock is taken.

It looks like we can't just use the `_watchingForExit` flag to check if `StopWatchingForExit` needs to be called because it's not updated in atomic way.

I've tried to find a way to reduce the cost of regression by making other things faster, but there are no low hanging fruits.

Name | Exc % | Exc | Inc % | Inc
-- | -- | -- | -- | --
coreclr!WKS::CFinalize::RegisterForFinalization | 14.8 | 44 | 14.8 | 44
coreclr!JIT_New | 13.1 | 39 | 61.7 | 183
coreclr!memset_repmovs | 9.8 | 29 | 9.8 | 29
coreclr!SVR::GCHeap::SetFinalizationRun | 9.1 | 27 | 9.1 | 27
coreclr!WKS::CFinalize::ScanForFinalization | 7.4 | 22 | 13.8 | 41
coreclr!JIT_MonReliableEnter_Portable | 7.1 | 21 | 7.1 | 21
coreclr!JIT_MonExit_Portable | 5.1 | 15 | 5.1 | 15
system.diagnostics.process.il!System.Diagnostics.Process.Close() | 5.1 | 15 | 18.2 | 54
coreclr!WKS::CFinalize::MoveItem | 3.7 | 11 | 3.7 | 11
coreclr!LazyMachStateCaptureState | 3.7 | 11 | 3.7 | 11
coreclr!GCInterface::SuppressFinalize | 3.4 | 10 | 13.1 | 39
system.diagnostics.process.il!Process.GetCurrentProcess | 2.7 | 8 | 74.1 | 220
coreclr!WKS::GCHeap::Alloc | 2.4 | 7 | 41.8 | 124

The most expensive part of `GetCurrentProcess` is allocating a finalizable instance of `Process` class. `Process` derives from `Component` which is finalizable, moreover it's not sealed and we simply have to pay 20% of time for that.

If we ever run into a situation where `GetCurrentProcessId().Id` is a performance bottleneck, we might consider making  `Process.GetCurrentProcessId()` public. This method is super fast and does not allocate anything.

**Comment by tmds** [comment]
> I think previously it was faster but incorrect 

Yes, we need to synchronize between `CompletionCallback` (which wants to emit the exited event) and `Close` (which wants to reset state).

> It looks like we can't just use the _watchingForExit flag to check if StopWatchingForExit needs to be called because it's not updated in atomic way.

I think this may be safe. When `_watchingForExit` is changed by the user, we'll see it because `Process` isn't thread-safe, so the user must use it in a thread-safe way. What we may not observe, is it being set to `false` from `CompletionCallback`. If we don't observe that, we'll still consider it to be `true` and call `StopWatchingForExit` (which will no-op).

**Comment by stephentoub** [comment]
> @stephentoub wdyt?

Sorry, @tmds, what change are you proposing?

**Comment by tmds** [comment]
@adamsitnik had a proposal but was doubting it is thread-safe:

> It looks like we can't just use the _watchingForExit flag to check if StopWatchingForExit needs to be called because it's not updated in atomic way.

I believe it's thread-safe because:

> When _watchingForExit is changed by the user, we'll see it because Process isn't thread-safe, so the user must use it in a thread-safe way. What we may not observe, is it being set to false from CompletionCallback. If we don't observe that, we'll still consider it to be true and call StopWatchingForExit (which will no-op).

**Comment by stephentoub** [comment]
Hmm, I'm not sure.  You might be right, but I'd need to walk through all the various ways you get here, as it seems we'd need to account for a variety of interactions.
